### PR TITLE
Fix/issue 1732 add payment methods

### LINKED ISSUE #1732: Use @wordpress/data in Add payment method modal
Use the data store from `@wordpress/data` to retrieve the list of enabled payment methods in the Add payment method modal.

Markup on 2021-05-05 at 18:19:30

The work here can be based on #1647.

## Additional matters

While working on this issue, answers to the following questions should also be included:

1. Should all available payment methods be displayed in the modal or just the ones that aren't yet enabled? Currently all methods are displayed but the designs seem to only display methods that are disabled. Maybe this should be changed?

Markup on 2021-05-05 at 18:24:20

2. Where in the list should the newly enabled methods be added? (At the end?)

### PR DESCRIPTION
**Pull request by brezocordero**
Fixes #1732

#### Changes proposed in this Pull Request

- This PR updates the `PaymentMethodsSelector` component to use the hook `useEnabledPaymentMethodIds`. The list of enabled payment methods is updated when the button `Add selected` is clicked.

- It also moves the modal inside of the component, same as for the [`Delete `button](https://github.com/Automattic/woocommerce-payments/pull/1790/files).

- Tests have been updated/added.
#### Testing instructions

- Ensure you have the "grouped settings" flag enabled in WCPay Dev Tools
- Visit http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout
- Click the "Add payment method" button
- The modal appears, currently enabled payment methods are checked.
- Cancelling the modal does not add new payment methods
- Selecting new methods and clicking on 'Add selected' adds the methods to the list.
- Remove a payment method by clicking on the trash icon.
- Click again on 'Add payment method' button
- Check the list is updated and the payment method removed is not checked.

There are [a couple of remaining questions ](https://github.com/Automattic/woocommerce-payments/issues/1732#issuecomment-842692186 ) that can be addressed on separate issues.

This issue does not cover clicking on 'Save settings' button on the settings screen.

-------------------

- [ ] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

### CODE REVIEW

File: client/payment-methods/index.js [hunk 1]
```diff
@@ -143,15 +129,7 @@ const PaymentMethods = () => {
...
-					<Button
-						className="payment-methods__add-payment-method"
-						onClick={ () =>
-							setPaymentMethodsSelectorModalVisible( true )
-						}
-						isSecondary
-					>
-						{ __( 'Add payment method', 'woocommerce-payments' ) }
-					</Button>
+					<PaymentMethodsSelector className="payment-methods__add-payment-method"></PaymentMethodsSelector>
```
Comment (frosso) [inline client/payment-methods/index.js:132]:
```suggestion
					<PaymentMethodsSelector className="payment-methods__add-payment-method" />
```

**Review by frosso — COMMENTED** [review]
Looks good - however, un-checking a payment method does not remove it from the list

**Comment by brezocordero** [comment]
> Looks good - however, un-checking a payment method does not remove it from the list

True, the delete method is not part of this component. 
I am going to implement [not listing the payment methods that are already active](https://github.com/Automattic/woocommerce-payments/issues/1732#issuecomment-843166463).
That will remove that case.

File: client/payment-methods/index.js (hunk 1 above)
Comment (brezocordero) [reply client/payment-methods/index.js]: Done!

**Comment by brezocordero** [comment]
@frosso 
Another round please :)
The selector only lists methods that haven't been enabled yet.

A couple of notes:
* I have extracted the hardcoded list of all available payment methods into an array and they are now displayed within a loop.
They could be passed as a prop or with a hook.

* I haven't implemented any logic when all payment methods are added. It could be [tackled on the issue you are working on](https://github.com/Automattic/woocommerce-payments/issues/1832) and do not display the whole section. We could also add a check inside of this component. I will let you decide.

File: client/settings/payment-methods-selector/index.js [hunk 2]
```diff
@@ -4,25 +4,54 @@
...
 
-const PaymentMethodsSelector = ( { onClose, enabledPaymentMethods = [] } ) => {
-	const [ paymentMethods, setPaymentMethods ] = useState(
-		enabledPaymentMethods.reduce( ( acc, value ) => {
-			acc[ value ] = true;
-			return acc;
-		}, {} )
-	);
+const PaymentMethodsSelector = ( { className } ) => {
+	const availablePaymentMethods = [
```
Comment (frosso) [inline client/settings/payment-methods-selector/index.js:20]: 👍 I have a PR that will solve this: https://github.com/Automattic/woocommerce-payments/pull/1848

File: client/settings/payment-methods-selector/test/index.js [hunk 3]
```diff
@@ -2,106 +2,200 @@
...
+		user.click( paymentMethodCheckbox );
 
-		const sepa = within( paymentMethods[ 2 ] );
-		expect( sepa.getByRole( 'checkbox' ).checked ).toBeFalsy();
+		expect( paymentMethodCheckbox ).toBeChecked();
+		expect(
+			useEnabledPaymentMethodIds().updateEnabledPaymentMethodIds
+		).not.toHaveBeenCalled();
+	} );
+	test( 'Selecting a payment method and clicking "Add selected" adds the method and closes the modal', () => {
```
Comment (frosso) [inline client/settings/payment-methods-selector/test/index.js:165]:
```suggestion
	} );

	test( 'Selecting a payment method and clicking "Add selected" adds the method and closes the modal', () => {
```

(minor) some breathing room between tests could help with readability (also in the previous lines)

**Review by frosso — APPROVED** [review]
New changes look good 👍

File: client/settings/payment-methods-selector/index.js [hunk 4]
```diff
@@ -4,25 +4,54 @@
...
+		enabledPaymentMethodIds: enabledMethodIds,
+		updateEnabledPaymentMethodIds: updateEnabledMethodIds,
+	} = useEnabledPaymentMethodIds();
+
+	const [
+		isPaymentMethodsSelectorModalOpen,
+		setIsPaymentMethodsSelectorModalOpen,
+	] = useState( false );
+
+	const [ paymentMethods, setPaymentMethods ] = useState();
```
[the remark this answers is not in the document]
Comment (brezocordero) [reply client/settings/payment-methods-selector/index.js]: Done!

File: client/settings/payment-methods-selector/index.js (hunk 2 above)
Comment (brezocordero) [reply client/settings/payment-methods-selector/index.js:20]: Very nice!

File: client/settings/payment-methods-selector/test/index.js (hunk 3 above)
Comment (brezocordero) [reply client/settings/payment-methods-selector/test/index.js]: Done!

**Pull request merged**